cranelift/isle/isle/src/trie_again.rs

```diff
@@ -0,0 +1,543 @@
+                    });
+                }
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(constraint);
+            }
+        }
+        Ok(())
+    }
+
+    /// Add `b` into the equivalence class of `a`. The caller must call this function a second time
+    /// to also add `a` into the equivalence class of `b`.
+    fn set_equal(&mut self, a: BindingId, b: BindingId) {
+        match self.equals.entry(a) {
+            Entry::Vacant(entry) => {
+                entry.insert(b);
+            }
+            Entry::Occupied(entry) => {
+                // This binding site was already part of an equivalence class. Find the element of
+                // the circular linked list that points back to this binding site and make it point
+                // to `b` instead.
+                let mut cur = *entry.get();
+                loop {
+                    let next = self.equals.get_mut(&cur).unwrap();
+                    if *next == a {
+                        *next = b;
+                        break;
+                    }
+                    cur = *next;
+                }
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+struct UnmatchableError {
+    pos: Pos,
+    constraint_a: Constraint,
+    constraint_b: Constraint,
+}
+
+#[derive(Debug, Default)]
+struct RuleSetBuilder {
+    current_rule: Rule,
+    binding_map: HashMap<Binding, BindingId>,
+    expr_map: HashMap<Expr, ExprId>,
+    unmatchable: Vec<UnmatchableError>,
+    rules: RuleSet,
+}
+
+impl RuleSetBuilder {
+    fn add_rule(
+        &mut self,
+        rule: &sema::Rule,
+        termenv: &sema::TermEnv,
+        tyenv: &sema::TypeEnv,
+        errors: &mut Vec<Error>,
+    ) {
+        self.current_rule.pos = rule.pos;
+        self.current_rule.prio = rule.prio;
+        self.current_rule.result = rule.visit(self, termenv);
+        self.normalize_equivalence_classes();
+        let rule = std::mem::take(&mut self.current_rule);
+
+        if self.unmatchable.is_empty() {
+            self.rules.rules.push(rule);
+        } else {
+            // If this rule can never match, drop it so it doesn't affect overlap checking.
+            errors.extend(self.unmatchable.drain(..).map(|err| {
+                let src = Source::new(
+                    tyenv.filenames[err.pos.file].clone(),
+                    tyenv.file_texts[err.pos.file].clone(),
+                );
+                Error::UnmatchableError {
+                    msg: format!(
+                        "rule requires binding to match both {:?} and {:?}",
+                        err.constraint_a, err.constraint_b
+                    ),
+                    src,
+                    span: Span::new_single(err.pos),
+                }
+            }))
+        }
+    }
+
+    // If a binding site is constrained and also required to be equal to another binding site, then
+    // copy the constraint and push the equality inside it.
```

Comment:
Can you expand this comment a bit more, explaining the overall scheme and why it works? Copy the constraints, meaning... push all constraints to one member of the equivalence class? Or...?

Reply:
And after reading the body of this function -- it's *extremely* confusing. Variable names are actively trying to melt my brain (`constraints` is a workqueue of sorts, `constraint` is a local thing; `binding` is something that comes from `constraints`, but `bindings` is something pulled out of `constraint`). It sort of reads like it's meant to normalize variant extractions.

I have to ask -- would it be better to take a more immutable approach to normalization here, where we (i) hashcons expression nodes, (ii) normalize all args (using e.g. a union-find) before creating a node on those args, and (iii) eagerly process equalities to create the union-find ahead of time (or as we go, at least)? Trying to reprocess post-facto here seems fraught with complexity and in any case is quite hard to understand (and thus maintain)!

Reply:
I originally tried to maintain an invariant during construction that every binding site could have a concrete constraint, or could participate in an equality constraint, but not both. I couldn't reason through making that work. My head hurt less when I just established that invariant after construction instead.

I've renamed many of the variables and added about 60 lines of comments to this function. Could you take another look and see if it makes more sense now?

Reply:
I've tried making the equality normalization incremental, in commit 0b9fe9040. It turns out to be pretty much exactly the same amount of code as doing it in a post-processing pass, unless I've missed a trick somewhere.

Also, I think it will do more work if three or more binding sites with enum-variant constraints get set equal to each other. In that case it runs n-1 times, and each time it builds all fields for two of the binding sites, instead of running once and building the fields for n binding sites all together. Then, if any field on the first binding site had a constraint, the same reasoning applies recursively.

So I'm thinking about dropping that change before merging. Would you take a look and see what you think?